File: python/providers/slack.py
"""
Slack Provider for commonlog
"""
import requests
from ..log_types import SendMethod, Provider

class SlackProvider(Provider):
    def send_to_channel(self, level, message, attachment, config, channel):
        original_channel = config.channel
        config.channel = channel
        self.send(level, message, attachment, config)
        config.channel = original_channel

    def send(self, level, message, attachment, config):
        formatted_message = self._format_message(message, attachment, config)
        if config.send_method == SendMethod.WEBCLIENT:
            self._send_slack_webclient(formatted_message, config)
        elif config.send_method == SendMethod.WEBHOOK:
            self._send_slack_webhook(formatted_message, config)
        else:
            raise ValueError(f"Unknown send method for Slack: {config.send_method}")

    def _format_message(self, message, attachment, config):
        formatted = ""

        # Add service and environment header
        if config.service_name and config.environment:
            formatted += f"*[{config.service_name} - {config.environment}]*\n"
        elif config.service_name:
            formatted += f"*[{config.service_name}]*\n"
        elif config.environment:
            formatted += f"*[{config.environment}]*\n"

        formatted += message

        if attachment and attachment.content:
            filename = attachment.file_name or "Trace Logs"
            formatted += f"\n\n*{filename}:*\n```\n{attachment.content}\n```"
        if attachment and attachment.url:
            formatted += f"\n\n*Attachment:* {attachment.url}"

        return formatted

    def _send_slack_webclient(self, formatted_message, config):
        # Use slack_token if available, otherwise fall back to token
        token = config.token
        if config.slack_token:
            token = config.slack_token
        
        url = "https://slack.com/api/chat.postMessage"
        headers = {"Authorization": f"Bearer {token}", "Content-Type": "application/json; charset=utf-8"}
        payload = {"channel": config.channel, "text": formatted_message}
        response = requests.post(url, headers=headers, json=payload)
        if response.status_code != 200:
            raise Exception(f"Slack WebClient response: {response.status_code}")

    def _send_slack_webhook(self, formatted_message, config):
        # For webhook, the token field contains the webhook URL
        webhook_url = config.token
        if not webhook_url:
            raise Exception("Webhook URL is required for Slack webhook method")
        
        payload = {"text": formatted_message}
        # If channel is specified, include it in the payload
        if config.channel:
            payload["channel"] = config.channel
        
        response = requests.post(webhook_url, json=payload)
        if response.status_code != 200:
            raise Exception(f"Slack webhook response: {response.status_code}")

File: go/providers/slack.go
package providers

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/alvianhanif/commonlog/go/types"
)

// SlackProvider implements Provider for Slack
type SlackProvider struct{}

func (p *SlackProvider) Send(level int, message string, attachment *types.Attachment, cfg types.Config) error {
	return p.SendToChannel(level, message, attachment, cfg, cfg.Channel)
}

func (p *SlackProvider) SendToChannel(level int, message string, attachment *types.Attachment, cfg types.Config, channel string) error {
	cfgCopy := cfg
	cfgCopy.Channel = channel
	switch cfgCopy.SendMethod {
	case types.MethodWebClient:
		return p.sendSlackWebClient(message, attachment, cfgCopy)
	case types.MethodWebhook:
		return p.sendSlackWebhook(message, attachment, cfgCopy)
	default:
		return fmt.Errorf("unknown send method for Slack: %s", cfgCopy.SendMethod)
	}
}

// formatMessage formats the alert message with optional attachment
func (p *SlackProvider) formatMessage(message string, attachment *types.Attachment, cfg types.Config) string {
	formatted := ""

	// Add service and environment header
	if cfg.ServiceName != "" && cfg.Environment != "" {
		formatted += fmt.Sprintf("*[%s - %s]*\n", cfg.ServiceName, cfg.Environment)
	} else if cfg.ServiceName != "" {
		formatted += fmt.Sprintf("*[%s]*\n", cfg.ServiceName)
	} else if cfg.Environment != "" {
		formatted += fmt.Sprintf("*[%s]*\n", cfg.Environment)
	}

	formatted += message

	if attachment != nil {
		if attachment.Content != "" {
			// Inline content - show as expandable code block
			filename := attachment.FileName
			if filename == "" {
				filename = "Trace Logs"
			}
			formatted += fmt.Sprintf("\n\n*%s:*\n```\n%s\n```", filename, attachment.Content)
		}
		if attachment.URL != "" {
			// External URL attachment
			formatted += fmt.Sprintf("\n\n*Attachment:* %s", attachment.URL)
		}
	}

	return formatted
}

func (p *SlackProvider) sendSlackWebhook(message string, attachment *types.Attachment, cfg types.Config) error {
	formattedMessage := p.formatMessage(message, attachment, cfg)

	// For webhook, the token field contains the webhook URL
	webhookURL := cfg.Token
	if webhookURL == "" {
		return fmt.Errorf("webhook URL is required for Slack webhook method")
	}

	payload := map[string]interface{}{
		"text": formattedMessage,
	}
	// If channel is specified, include it in the payload
	if cfg.Channel != "" {
		payload["channel"] = cfg.Channel
	}

	data, _ := json.Marshal(payload)
	fmt.Printf("[SlackProvider] Sending webhook to URL: %s, payload: %s\n", webhookURL, string(data))
	req, _ := http.NewRequest("POST", webhookURL, bytes.NewBuffer(data))
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Printf("[SlackProvider] Error sending webhook request: %v\n", err)
		return err
	}
	defer resp.Body.Close()

	// Log response data
	respData := new(bytes.Buffer)
	respData.ReadFrom(resp.Body)
	fmt.Printf("[SlackProvider] Slack webhook response status: %d, body: %s\n", resp.StatusCode, respData.String())

	if resp.StatusCode != 200 {
		return fmt.Errorf("slack webhook response: %d", resp.StatusCode)
	}
	fmt.Println("[SlackProvider] Webhook sent successfully")
	return nil
}

func (p *SlackProvider) sendSlackWebClient(message string, attachment *types.Attachment, cfg types.Config) error {
	formattedMessage := p.formatMessage(message, attachment, cfg)

	// Use SlackToken if available, otherwise fall back to Token
	token := cfg.Token
	if cfg.SlackToken != "" {
		token = cfg.SlackToken
	}

	url := "https://slack.com/api/chat.postMessage"
	headers := map[string]string{"Authorization": "Bearer " + token, "Content-Type": "application/json; charset=utf-8"}
	payload := map[string]interface{}{
		"channel": cfg.Channel,
		"text":    formattedMessage,
	}
	data, _ := json.Marshal(payload)
	fmt.Printf("[SlackProvider] Sending to channel: %s, payload: %s\n", cfg.Channel, string(data))
	req, _ := http.NewRequest("POST", url, bytes.NewBuffer(data))
	for k, v := range headers {
		req.Header.Set(k, v)
	}
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		fmt.Printf("[SlackProvider] Error sending request: %v\n", err)
		return err
	}
	defer resp.Body.Close()

	// Log response data
	respData := new(bytes.Buffer)
	respData.ReadFrom(resp.Body)
	fmt.Printf("[SlackProvider] Slack WebClient response status: %d, body: %s\n", resp.StatusCode, respData.String())

	if resp.StatusCode != 200 {
		return fmt.Errorf("slack WebClient response: %d", resp.StatusCode)
	}
	fmt.Println("[SlackProvider] Message sent successfully")
	return nil
}
